### PR TITLE
feat: 🎸 Migration from Magic to Clerk

### PR DESCRIPTION
**Pull request by hamzabell**

### CODE REVIEW

File: package.json
Comment (janhesters) [inline package.json:1]: We should be able to uninstall al the magic packages, right?

File: prisma/schema.prisma [hunk 1]
```diff
@@ -26,6 +23,8 @@ model UserProfile {
   notes                      Note[]
   createdInviteLinks         OrganizationInviteLink[]
   inviteLinkUse              InviteLinkUse[]
+  clerkId                    String                          @unique
+  hasFalseClerkId            Boolean                         @default(false)
```
Comment (janhesters) [inline prisma/schema.prisma:27]: What is this for?

File: app/routes/login.test.ts
Comment (janhesters) [inline app/routes/login.test.ts:1]: We need to replace all tests instead of deleting them.

File: app/routes/register.test.ts
Comment (janhesters) [inline app/routes/register.test.ts:1]: Same here, the actions still need to be thoroughly tested.

File: app/features/user-profile/user-profile-factories.server.ts [hunk 2]
```diff
@@ -16,15 +16,17 @@ export const createUserProfile: Factory<UserProfile> = ({
   id = '',
   createdAt = new Date(),
   updatedAt = new Date(),
-  did = '',
+  clerkId = '',
   email = '',
   name = '',
+  hasFalseClerkId = false,
```
Comment (janhesters) [inline app/features/user-profile/user-profile-factories.server.ts:22]: Do we need this? What is this for?

File: app/features/user-profile/user-profile-model.server.ts [hunk 3]
```diff
@@ -145,3 +150,13 @@ export async function updateUserProfileInDatabaseById({
 export async function deleteUserProfileFromDatabaseById(id: UserProfile['id']) {
   return prisma.userProfile.delete({ where: { id } });
 }
+
+export async function deleteUserProfileFromDatabaseByClerkUserId(
```
Comment (janhesters) [inline app/features/user-profile/user-profile-model.server.ts:154]: Let's add TSDoc.

File: app/https.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import { redirect } from '@remix-run/node';
+
+export const enforceHttps = (request: Request) => {
```
Comment (janhesters) [inline app/https.ts:3]: Should we add TSDoc?

File: app/https.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  let url = new URL(request.url);
+  let hostname = url.hostname;
+  let proto = request.headers.get('X-Forwarded-Proto') ?? url.protocol;
```
Comment (janhesters) [inline app/https.ts:6]: Neither hostname nor proto change. Should they be `const`? Also should proto be named `protocol`?

File: app/root.tsx [hunk 6]
```diff
@@ -7,18 +9,17 @@ import { json } from '@remix-run/node';
 import {
   isRouteErrorResponse as getIsRouteErrorResponse,
   Links,
+  LiveReload,
```
Comment (janhesters) [inline app/root.tsx:12]: Don't think we need the `LiveReload` tag with Vite, do we?

File: app/root.tsx [hunk 7]
```diff
@@ -98,58 +106,62 @@ function useChangeLanguage(locale: string) {
 }
 
 function App() {
-  return <Outlet />;
-}
-
-export function Layout({ children }: { children: ReactNode }) {
```
Comment (janhesters) [inline app/root.tsx:104]: Why did we get rid of the `<Layout />`? And then manually introduced it back to both the error and the app?

File: app/routes/login.tsx [hunk 8]
```diff
@@ -63,66 +62,66 @@ export default function Login() {
...
+
+        const login = await signIn.create({
+          identifier: email,
+          redirectUrl,
+        });
+
+        const emailLink = signIn.createEmailLinkFlow();
+        setIsVerifyingEmail(true);
+        const emailVerified = await emailLink.startEmailLinkFlow({
+          // @ts-ignore
```
Comment (janhesters) [inline app/routes/login.tsx:89]: Why are we ignoring here? Maybe better use `@ts-expect-error` with a reason?

File: app/routes/login.tsx [hunk 9]
```diff
@@ -134,57 +133,70 @@ export default function Login() {
...
+                      <Loader2Icon className="mr-2 size-4 animate-spin" />
+                      {t('authenticating')}
+                    </>
+                  ) : (
+                    t('login')
+                  )}
+                </Button>
+              </fieldset>
+            </Form>
+          </FormProvider>
```
Comment (janhesters) [inline app/routes/login.tsx:188]: I think a ternary for these two could be easier to read.

File: app/routes/register.tsx [hunk 10]
```diff
@@ -65,66 +64,60 @@ export default function Register() {
...
-  useEffectOnce(() => {
-    downloadMagicStaticAssets().catch(() => {
-      form.setError('email', { message: 'register:failed-to-load-magic' });
-    });
-  });
+  const navigation = useNavigation();
+  const isRegisteringViaEmail =
+    navigation.formData?.get('intent') === 'emailRegistration' ||
+    navigation.formData?.get('intent') === 'clerkEmailRegistration' ||
+    isRegisteringWithClerk;
```
Comment (janhesters) [inline app/routes/register.tsx:80]: NIT: Should we put these below the `useEffect`?

File: app/features/user-authentication/awaiting-email-confirmation.tsx [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+          <TextLink to={`/login${token ? `?token=${token}` : ''}`}>
+            {t('log-in-to-your-account')}
+          </TextLink>
+        </Text>
+      </div>
+    </div>
+  );
+};
+export const AwaitingLoginEmailVerification = ({
+  onCancel,
```
Comment (janhesters) [inline app/features/user-authentication/awaiting-email-confirmation.tsx:48]: Looks like we never use `onCancel` here 🤔 Why do we have it?

File: app/routes/register.tsx [hunk 12]
```diff
@@ -135,100 +128,117 @@ export default function Register() {
...
+            cancelEmailVerificationRef.current?.();
+            setIsVerifyingEmail(false);
+            setIsRegisteringWithClerk(false);
+          }}
+          email={actionData?.email || 'tobiz2002@gmail.com'}
+          token={loaderData.token}
+        />
+      )}
+
+      {!isVerifyingEmail && (
```
Comment (janhesters) [inline app/routes/register.tsx:143]: Should we use a ternary here?

File: app/features/user-authentication/client/user-authentication-actions.server.ts [hunk 13]
```diff
@@ -321,6 +340,8 @@ export const registerAction = async ({
             : undefined,
         });
       } catch {
+        await deleteUserProfileFromDatabaseByClerkUserId(clerkUserId);
```
[the remark this answers is not in the document]
Comment (hamzabell) [reply app/features/user-authentication/client/user-authentication-actions.server.ts:343]: The reason for this is if there occurs an error during the creation of the profile, the user still using the failing email should be able to retry, one issue I remembered facing  with ACOS is that without that line, there tends to be a conflict if after registration failed you try again, but the app sometimes has already created the account

File: app/test/test-utils.ts [hunk 14]
```diff
@@ -50,6 +49,15 @@ export async function setupUserWithOrgAndAddAsMember({
...
 }
 
+/**
+ * Saves the user profile and organization to the database and adds the user as
+ * a member of the organization with a auth session id.
+ *
+ * @param options - Optional parameter containing the organization and user
+ * objects to be saved.
+ * @returns - An object containing the saved organization and user.
+ */
```
Comment (janhesters) [inline app/test/test-utils.ts:59]: What's going on here? 😄

File: playwright.config.ts [hunk 15]
```diff
@@ -21,6 +21,13 @@ export default defineConfig({
   workers: 1,
   /* Reporter to use. See https://playwright.dev/docs/test-reporters */
   reporter: 'html',
+  webServer: {
+    command: 'npm run dev',
+    url: process.env.BASE_URL || 'http://localhost:3000',
+    reuseExistingServer: true,
+    stdout: 'ignore',
+    stderr: 'pipe',
+  },
```
Comment (janhesters) [inline playwright.config.ts:30]: This looks like, it should be removed, and the commented out code should go back in.

File: playwright/e2e/user-authentication/login.spec.ts
Comment (janhesters) [inline playwright/e2e/user-authentication/login.spec.ts:1]: Do we really want to get rid of the login happy path test?

**Review by janhesters — APPROVED** [review]

**Pull request merged**